What should we use as the eth-trie hash type?

Two years ago, I made a choice to represent the hash as a keccak::H256. It's been long enough that the choice might not be the best one anymore (it was just a byte slice before, so it's still probably better than that)

When trying out v0.2.0, which includes this change, in trin, that seems like the biggest breaking change. So, what is the best option here?

One option would be to switch the type from the keccak one to the `ethereum_types` `H256`. That requires exposing ethereum_types as a proper dependency (not just a dev one). Also, it doesn't fully fix the issue:

```sh
error[E0277]: the trait bound `ethereum_types::H256: ssz::Decode` is not satisfied
  --> ethportal-api/src/types/consensus/body.rs:11:59
   |
11 | #[derive(Debug, PartialEq, Clone, Deserialize, Serialize, Decode, Encode)]
   |                                                           ^^^^^^ the trait `ssz::Decode` is not implemented for `ethereum_types::H256`
   |
   = help: the following other types implement trait `ssz::Decode`:
             (A, B)
             (A, B, C)
             (A, B, C, D)
             (A, B, C, D, E)
             (A, B, C, D, E, F)
             (A, B, C, D, E, F, G)
             (A, B, C, D, E, F, G, H)
             (A, B, C, D, E, F, G, H, I)
           and 98 others
   = note: this error originates in the derive macro `Decode` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `ethereum_types::H256: ssz::Decode` is not satisfied
   --> ethportal-api/src/types/consensus/body.rs:15:19
    |
15  |     pub graffiti: H256,
    |                   ^^^^ the trait `ssz::Decode` is not implemented for `ethereum_types::H256`
    |
    = help: the following other types implement trait `ssz::Decode`:
              (A, B)
              (A, B, C)
              (A, B, C, D)
              (A, B, C, D, E)
              (A, B, C, D, E, F)
              (A, B, C, D, E, F, G)
              (A, B, C, D, E, F, G, H)
              (A, B, C, D, E, F, G, H, I)
            and 98 others
note: required by a bound in `SszDecoderBuilder::<'a>::register_type`
   --> /home/carver/.cargo/registry/src/index.crates.io-6f17d22bba15001f/eth2_ssz-0.4.0/src/decode.rs:176:29
    |
176 |     pub fn register_type<T: Decode>(&mut self) -> Result<(), DecodeError> {
    |                             ^^^^^^ required by this bound in `SszDecoderBuilder::<'a>::register_type`
```

The most obvious options are:
- keep keccak:H256 as the exposed type, dependent projects just have to adjust (and maybe have the same encoding issues?)
- switch to ethereum_types:H256, and try to get the ssz lib to write encode/decode for this type
- switch it back to a bytes slice, and lose the guarantee of it being a 32-byte slice

I'll sleep on it. Open to discussion!
I guess we could try to get the ssz library to implement Encode and Decode for the H256 type

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,6 +12,7 @@ homepage = "https://github.com/carver/eth-trie.rs"
 documentation = "https://docs.rs/eth_trie"
 
 [dependencies]
+ethereum-types = "0.12.1"
 hashbrown = "0.14.0"
 keccak-hash = "0.10.0"
 log = "0.4.16"
@@ -22,7 +23,6 @@ rlp = "0.5.1"
 rand = "0.8.3"
 hex = "0.4.2"
 criterion = "0.5.1"
-ethereum-types = "0.14.1"
 uuid = { version = "1.4.1", features = ["serde", "v4"] }
 
 [[bench]]

--- a/src/errors.rs
+++ b/src/errors.rs
@@ -1,7 +1,7 @@
 use std::error::Error;
 use std::fmt;
 
-use keccak_hash::H256;
+use ethereum_types::H256;
 use rlp::DecoderError;
 
 use crate::nibbles::Nibbles;

--- a/src/node.rs
+++ b/src/node.rs
@@ -1,6 +1,6 @@
 use std::sync::{Arc, RwLock};
 
-use keccak_hash::H256;
+use ethereum_types::H256;
 
 use crate::nibbles::Nibbles;
 

--- a/src/trie.rs
+++ b/src/trie.rs
@@ -1,8 +1,8 @@
-use keccak_hash::KECCAK_NULL_RLP;
 use std::sync::{Arc, RwLock};
 
+use ethereum_types::H256;
 use hashbrown::{HashMap, HashSet};
-use keccak_hash::{keccak, H256};
+use keccak_hash::{keccak, KECCAK_NULL_RLP};
 use log::warn;
 use rlp::{Prototype, Rlp, RlpStream};
 
@@ -225,7 +225,7 @@ where
     pub fn new(db: Arc<D>) -> Self {
         Self {
             root: Node::Empty,
-            root_hash: KECCAK_NULL_RLP,
+            root_hash: KECCAK_NULL_RLP.as_fixed_bytes().into(),
 
             cache: HashMap::new(),
             passing_keys: HashSet::new(),
@@ -388,7 +388,7 @@ where
     ) -> TrieResult<Option<Vec<u8>>> {
         let proof_db = Arc::new(MemoryDB::new(true));
         for node_encoded in proof.into_iter() {
-            let hash = keccak(&node_encoded);
+            let hash: H256 = keccak(&node_encoded).as_fixed_bytes().into();
 
             if root_hash.eq(&hash) || node_encoded.len() >= HASHED_LENGTH {
                 proof_db.insert(hash.as_bytes(), node_encoded).unwrap();
@@ -764,7 +764,7 @@ where
         let root_hash = match self.write_node(&self.root.clone()) {
             EncodedNode::Hash(hash) => hash,
             EncodedNode::Inline(encoded) => {
-                let hash = keccak(&encoded);
+                let hash: H256 = keccak(&encoded).as_fixed_bytes().into();
                 self.cache.insert(hash.as_bytes().to_vec(), encoded);
                 hash
             }
@@ -813,7 +813,7 @@ where
         if data.len() < HASHED_LENGTH {
             EncodedNode::Inline(data)
         } else {
-            let hash = keccak(&data);
+            let hash: H256 = keccak(&data).as_fixed_bytes().into();
             self.cache.insert(hash.as_bytes().to_vec(), data);
 
             self.gen_keys.insert(hash.as_bytes().to_vec());
@@ -931,7 +931,8 @@ mod tests {
     use std::collections::{HashMap, HashSet};
     use std::sync::Arc;
 
-    use keccak_hash::{H256, KECCAK_NULL_RLP};
+    use ethereum_types::H256;
+    use keccak_hash::KECCAK_NULL_RLP;
 
     use super::{EthTrie, Trie};
     use crate::db::{MemoryDB, DB};

